### ISSUE TITLE
Conflict between `no-fallthrough` rule and `reportUnusedDisableDirectives` option

### ISSUE THREAD

**Issue by jens-duttke** [eslint/eslint#13627]
**Tell us about your environment**

* **ESLint Version:** 7.7.0
* **Node Version:** 14.8.0
* **npm Version:** 6.14.7

**What parser (default, `@babel/eslint-parser`, `@typescript-eslint/parser`, etc.) are you using?**

default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
	ignorePatterns: [
		'node_modules/'
	],
	parserOptions: {
		ecmaVersion: 2020,
		sourceType: 'module'
	},
	env: {
		browser: true,
		commonjs: true
	},
	reportUnusedDisableDirectives: true,
	rules: {
		'no-fallthrough': 'error'
	}
};
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
const type = 'bar';
let xyz = 'x';

switch (type) {
	case 'foo':
	case 'bar':
		xyz += 'y';

	// eslint-disable-next-line no-fallthrough
	case 'baz':
		xyz += 'z';
		break;

	default:
}

console.log(xyz);
```

```bash
eslint .
```

**What did you expect to happen?**
**What actually happened? Please include the actual, raw output from ESLint.**

Without `// eslint-disable-next-line no-fallthrough` Eslint reports `Expected a 'break' statement before 'case'`, which is absolutely correct.
But adding this line, ESLint reports `Unused eslint-disable directive (no problems were reported from 'no-fallthrough')`, which is not correct, because this directive suppresses the `no-fallthrough` error.

I've created a minimal test setup here:
https://github.com/jens-duttke/eslint-bug-test

**Are you willing to submit a pull request to fix this bug?**

No

**Comment by anikethsaha** [comment]
Thanks for the report. 

This is a bug. old discussion https://github.com/eslint/eslint/issues/13260

here is the details: https://github.com/eslint/eslint/issues/13260#issuecomment-624114328

**Comment by mdjermanovic** [comment]
Maybe this isn't technically a bug, but it would be good to somehow avoid this coincidence.

A workaround for users who don't want to use rule's comments is to configure `commentPattern` to something that cannot be matched, like `commentPattern: "^.^"`. That would effectively disable this option and thus enforce the use of `eslint-disable` comments where needed.

Aside from that, how can we improve this?

Some ideas:

* Ignore `eslint` comments. This needs some further analysis. For example, if the last comment is an `eslint` comment, should we check the comment before that one? Also, this might be a breaking change. 
* Remove default pattern. I think it would be better if this rule didn't have a default pattern, but removing it now looks like a big breaking change.
* Add an official way to disable the `commentPattern` feature.

**Comment by anikethsaha** [comment]
> Ignore eslint comments. This needs some further analysis. For example, if the last comment is an eslint comment, should we check the comment before that one? Also, this might be a breaking change.

This seems much safer. But yea breaking change. 

> * Add an official way to disable the `commentPattern` feature.

if it's about not having any patterns. then having `{ commentPattern: '.' }` might work ? [demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tZmFsbHRocm91Z2g6IFtcImVycm9yXCIsIHsgXCJjb21tZW50UGF0dGVyblwiOiBcIi5cIiB9XSovXG5cbnN3aXRjaChmb28pIHtcbiAgICBjYXNlIDE6XG4gICAgICAgIGRvU29tZXRoaW5nKCk7XG4gICAgICAgIC8vIGJyZWFrIG9taXR0ZWRcblxuICAgIGNhc2UgMjpcbiAgICAgICAgZG9Tb21ldGhpbmcoKTtcbn1cblxuc3dpdGNoKGZvbykge1xuICAgIGNhc2UgMTpcbiAgICAgICAgZG9Tb21ldGhpbmcoKTtcbiAgICAgICAgLy8gY2F1dGlvbjogYnJlYWsgaXMgb21pdHRlZCBpbnRlbnRpb25hbGx5XG5cbiAgICBkZWZhdWx0OlxuICAgICAgICBkb1NvbWV0aGluZygpO1xufSIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6Niwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6eyJqc3giOnRydWV9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)
cause the option is overriding the default.

Or we can fix the default comment pattern to not include eslint directives ! (maybe breaking change as well)

**Comment by mdjermanovic** [comment]
> if it's about not having any patterns. then having `{ commentPattern: '.' }` might work ? [demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tZmFsbHRocm91Z2g6IFtcImVycm9yXCIsIHsgXCJjb21tZW50UGF0dGVyblwiOiBcIi5cIiB9XSovXG5cbnN3aXRjaChmb28pIHtcbiAgICBjYXNlIDE6XG4gICAgICAgIGRvU29tZXRoaW5nKCk7XG4gICAgICAgIC8vIGJyZWFrIG9taXR0ZWRcblxuICAgIGNhc2UgMjpcbiAgICAgICAgZG9Tb21ldGhpbmcoKTtcbn1cblxuc3dpdGNoKGZvbykge1xuICAgIGNhc2UgMTpcbiAgICAgICAgZG9Tb21ldGhpbmcoKTtcbiAgICAgICAgLy8gY2F1dGlvbjogYnJlYWsgaXMgb21pdHRlZCBpbnRlbnRpb25hbGx5XG5cbiAgICBkZWZhdWx0OlxuICAgICAgICBkb1NvbWV0aGluZygpO1xufSIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6Niwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6eyJqc3giOnRydWV9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)
> cause the option is overriding the default.

I think we need the opposite if we want to allow users to enforce the use of "standard" `eslint-disable` comments: a documented way to disable the comment feature of this rule, like `commentPattern: null` or `commentPattern: false`.

**Comment by anikethsaha** [comment]
I think having `null` or `false` would still use the default pattern so the bug mentioned https://github.com/eslint/eslint/issues/13260 will remain there. 
I added `['.']` in order to allow all. basically no checking for the pattern. 

WDYT ?

**Comment by mdjermanovic** [comment]
`"."` makes all non-empty comments being the comments that allow fall-through, so I don't think it helps to avoid conflicts with `eslint-disable`.

For example, try to run eslint with `--report-unused-disable-directives` on the following code:

```js
/* eslint no-fallthrough: ["error", { commentPattern: "." }] */

switch (type) {
	case 'foo':
	case 'bar':
		xyz += 'y';

	// eslint-disable-next-line no-fallthrough
	case 'baz':
		xyz += 'z';
		break;

	default:
}
```

```
  8:2  error  Unused eslint-disable directive (no problems were reported from 'no-fallthrough')
```

It's still the same issue. We just replaced one pattern (the default one) that matches `"eslint-disable-next-line no-fallthrough"` with another pattern that also matches.

`null` and `false` currently produce schema validation errors, so we can use one of them as a value that disables the "intentional comment" feature of this rule. That still wouldn't help to avoid confusion with the default behavior of this rule (changing or removing the default pattern would help, but I'm not sure should we do that even in a major version), but at least provides a clear way to disable this feature. A workaround that can be used at the moment is to set a pattern that cannot match anything, like `{ commentPattern: "[]" }`.

**Comment by mdjermanovic** [comment]
#13746 proposal for `"commentPattern": null`
